### ISSUE TITLE
when field it not required and value is empty

### ISSUE THREAD

**Issue by er314** [SandroMiguel/verum-php#11]
Hi, 

Question : When a field is NOT required, and the value is EMPTY, what should be the result of its validation ?

-> My opinion : Whatever the rules applied on this field, it should ALWAYS PASS. 

In other words, my view on the desired behaviour in terms of WHEN the validation checks should be performed :
. principle A. : If a field is required, then it must have a non-empty value
. + principle B. : If a field has a non-empty value (required or not), then any rule configured on it must be checked ; otherwise no check and PASS

principle A. :
	-> this is implemented in the Required rule + in the RequiredIf rule
	-> if required and value is empty -> NO-PASS
principle B. :
	-> this means that ALL Rules shall just return true if the field is empty

Examples : 

example 1.a : a numeric field, with rule min = 10 ; value empty -> is should pass       
example 1.b : but if the field is required -> it should fail
-> for instance, a field "your age", not-required, type number, min=1, max=99 -> if user doesn't submit a value, it should pass

example 2.a : a textual field, with rule min_length = 10 ; value empty -> is should pass
example 2.b : but if the field is required -> it should fail
-> for instance, a field "your favorite sport", not-required, type text, min_length=3 -> if user doesn't submit a value, it should pass

example 3.a : a text field, with a regex pattern ; value empty -> it should pass
example 3.b : but if the field is required -> it should pass or fail, depending if the pattern allows no chars or not
-> for instance, a field "product ID", not-required, type text, regex = 3 blocks of 4 chars separated by hyphen -> if user doesn't submit a value, it should pass

You get the point...

Now, does verum-php have this behavior ?
Well, I'm writing this issue because no, it doesn't.
On the 3 examples above, 
- 1.a. :  -> pass
- 2.a. :  -> no-pass
- 3.a. :  -> no-pass

And in the code, we can see that :
- 1.a. is ok  because Numeric.php is "compliant with principle B." : if ($this->fieldValue === '') { return true; }
- 2.a. is nok because MinLength.php doesn't do what Numeric.php does when fieldValue is empty
- 3.a. is nok because Regex.php doesn't do what Numeric.php does when fieldValue is empty

Sooo.... what's your opinion on all of this ? :-)

(note : it's not applicable to the special rule Equals, which is just fine as-is)

**Comment by SandroMiguel** [comment]
Hi @er314,
I'm reviewing all the rules and I'm going to create a table for each rule like this:
| Value       | alpha              | alpha + required   |
| ----------- | ------------------ | ------------------ |
| null        | :heavy_check_mark: | :x:                |
| ''          | :heavy_check_mark: | :x:                |
| '0'         | :x:                | :x:                |
| 0           | :x:                | :x:                |
| false       | :x:                | :x:                |
| []          | :x:                | :x:                |
| -1          | :x:                | :x:                |
| 1           | :x:                | :x:                |
| true        | :x:                | :x:                |
| 'some text' | :heavy_check_mark: | :heavy_check_mark: |

I'm working on this branch -> https://github.com/SandroMiguel/verum-php/tree/t-358/add-rule-tables
As soon as I finish I will tell you.

**Comment by er314** [comment]
Hi, 

Yes, not trivial...
I agree that this approach looks good.
Attention should be paid to the articulation along with the behavior of the "required*" rules.

Here is how I see things, based on your example for the "alpha" rule :

The conditions your defined in the table for the "alpha" rule fully makes sense, when articulated as follows : 

- "alpha" rule : 

		.1. pass when no value 
			= pass when value is (null or empty string)
		.2. or, pass when the logic of the constraint is respected
			= pass when ctype_alpha is true

- "required*" rules :

		.3. pass when there is a value
			= pass when NOT 1.
			= pass when value is NOT (null or empty string)
			
The difficulty is that the "required*" rules are (currently) "agnostic" wrt to the other rules, so they must ensure that :
	- whatever the other rules, the "pass when NOT 1." is satisfied.
	- also, without any other rules applied on a field, the "required" rule must be self-sufficient.

So, for avoiding trouble and for avoiding adding more complexity, we can **hope** that the **conclusion for all the rules** conditions tables is :

		.1. is always = pass when value is (null or empty string)
		.2. is rule-specific
		consequently,
		.3. can always be = pass when value is NOT (null or empty string)


I don't know in advance if this assumption is valid for all rules.
On a few examples, it seems valid :

	- "contains" rule :
		.1. pass when no value 
			= pass when value is (null or empty string)
		.2. or, pass when the logic of the constraint is respected
			= pass when in_array(value, arrayOfValidValues)
	- "required*" rules :
		.3. pass when there is a value
			= pass when NOT 1.
			= pass when value is NOT (null or empty string)    -> OK, still consistent with "required*" logic when applied to "alpha" rule

	- "numeric" rule :
		.1. pass when no value 
			= pass when value is (null or empty string)
		.2. or, pass when the logic of the constraint is respected
			= pass when is_numeric
	- "required*" rules :
		.3. pass when there is a value
			= pass when NOT 1.
			= pass when value is NOT (null or empty string)    -> OK, still consistent with "required*" logic when applied to "alpha" rule / to "contains" rule	

	- "booleanValue" rule :
		.1. pass when no value 
			= pass when value is (null or empty string)
		.2. or, pass when the logic of the constraint is respected
			= pass when is_bool
	- "required*" rules :
		.3. pass when there is a value
			= pass when NOT 1.
			= pass when value is NOT (null or empty string)    -> OK, still consistent with "required*" logic when applied to "alpha" rule / to "numeric" rule / to "contains" rule
	-> Here, I can see that your current code is more advanced for boolean check : you make boolean-ish validation.
	This makes the "required*" rules more complex and inconsistent with when applied to other rules.
	Maybe, for the sake of simplicity, this boolean-ish behavior may be abandonned ? The calling code can always perform boolean-ish validation with e.g. contains => [0, '0', 1, '1', false, true, "off", "on"]

If the above assumption proves to be valid for all rules, then bingo :-) , the result would be :
	- a clear definition of each rule validation conditions, 
	- a clear definition of the "required*" validation conditions : pass when value is NOT (null or empty string)

Lastly, cherry on the cake :-) :
Regarding the "required*" rules, for making them as "universal" as possible (ie. for making them plenty useful even without any other rules), it is tempting to 
not only do this :
`		pass when value is NOT (null or empty string)`
but also do this :
`		pass when value is NOT (null or empty string or empty array)`
-> Is it feasible, without breaking any other rule ?
In other words, would it break any .2. logic conditions for any other rules ?
-> TBD with all rules, but I think that most probably it's feasible, ie. most probably there's no rule for which the logic of the constraint allows empty array.

Ok, this is not a matured thought, so take it with a grain of salt ! :-)

**Comment by SandroMiguel** [comment]
Hey @er314, I just released Verum PHP 2.0 and added tables to all rules in readme.md.

**Comment by er314** [comment]
Hi, thanks for the information ! I'll have a look into it later on.
Now I close this issue, as it's no more relevant.
